klog/klog.c

```diff
@@ -0,0 +1,354 @@
+
+	hal_memset(r, 0, sizeof(klog_reader_t));
+
+	r->nonblocking = !!nonblocking;
+	r->pid = pid;
+
+	proc_lockSet(&klog_common.lock);
+	r->ridx = klog_common.head;
+	LIST_ADD(&klog_common.readers, r);
+	proc_lockClear(&klog_common.lock);
+
+	return 0;
+}
+
+
+static ssize_t _klog_readln(klog_reader_t *r, char *buf, size_t sz)
+{
+	ssize_t n = 0;
+
+	while (r->ridx < klog_common.tail && n < sz) {
+		buf[n++] = _fifo_get(r->ridx++);
+		if (buf[n - 1] == '\n' || buf[n - 1] == '\0')
+			break;
+	}
+
+	/* Always end with newline */
+	if (n > 0 && buf[n - 1] != '\n') {
+		if (buf[n - 1] == '\0') {
+			buf[n - 1] = '\n';
+		}
+		else if (n < sz) {
+			buf[n++] = '\n';
+		}
+		else {
+			buf[n - 1] = '\n';
+			r->ridx--;
+		}
+	}
+
+	return n;
+}
+
+
+static ssize_t klog_read(klog_reader_t *r, char *buf, size_t sz)
+{
+	ssize_t ret;
+
+	proc_lockSet(&klog_common.lock);
+	/* We need to catch up the ring buffer's head */
+	if (r->ridx < klog_common.head)
+		r->ridx = klog_common.head;
+
+	ret = _klog_readln(r, buf, sz);
+
+	proc_lockClear(&klog_common.lock);
+
+	return ret;
+}
+
+
+static void _klog_updateReaders(void)
+{
+	klog_reader_t *r = klog_common.readers;
+	msg_t msg;
+
+	msg.type = mtRead;
+
+	if (r != NULL) {
+		do {
+			if (r->waiting) {
+				msg.pid = r->pid;
+				msg.o.data = r->odata;
+				msg.o.size = r->osize;
+				msg.o.io.err = _klog_readln(r, msg.o.data, msg.o.size);
+				r->waiting = 0;
+				proc_respond(0, &msg, r->rid);
+			}
+			r = r->next;
+		} while (r != klog_common.readers);
+	}
+}
+
+
+static void klog_close(unsigned pid)
+{
+	klog_reader_t *r;
+	msg_t msg;
+
+	msg.type = mtRead;
+	proc_lockSet(&klog_common.lock);
+	if ((r = klog_readerFind(pid)) != NULL) {
+		if (r->waiting) {
+			msg.pid = r->pid;
+			msg.o.data = r->odata;
+			msg.o.size = r->osize;
+			msg.o.io.err = -EIO;
+			proc_respond(klog_common.port, &msg, r->rid);
+		}
+		LIST_REMOVE(&klog_common.readers, r);
+		vm_kfree(r);
+	}
+	proc_lockClear(&klog_common.lock);
+}
+
+
+static int klog_devctl(msg_t *msg)
+{
+	ioctl_in_t *in;
+	ioctl_out_t *out;
+
+	in = (ioctl_in_t *)msg->i.raw;
+	out = (ioctl_out_t *)msg->o.raw;
+	/*
+	 * We need to handle isatty(), which
+	 * only checks if a device responds to TCGETS
+	 */
+	if (in->request == TCGETS) {
+		out->err = EOK;
+	}
+	else {
+		out->err = -EINVAL;
+	}
+
+	return 0;
+}
+
+
+int klog_write(const char *data, size_t len)
+{
+#if KLOG_ENABLE
+	int i = 0, overwrite = 0;
+	char c;
+
+	if (len == 0)
+		return 0;
+
+	proc_lockSet(&klog_common.lock);
+
+	if (klog_common.tail + len >= klog_common.head + KLOG_BUFSZ)
+		overwrite = 1;
+
+	while (i < len)
+		_fifo_push(data[i++]);
+
+	if (overwrite) {
+		do {
+			c = _fifo_pop();
+		} while (c != '\n' && c != '\0' && !_fifo_empty());
+	}
+
+	_klog_updateReaders();
+	proc_lockClear(&klog_common.lock);
+#else
+	hal_consolePrint(ATTR_NORMAL, data);
+#endif /* KLOG_ENABLE */
+
+	return len;
+}
+
+
+static void klog_readerBlock(klog_reader_t *r, msg_t *msg, unsigned long rid)
```

Comment:
Edge case is not handled - what if we do two blocking reads on one fd?
We need a list of read msg within a list of readers, I'm afraid